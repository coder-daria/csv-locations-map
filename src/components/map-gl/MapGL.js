import React, { useState, memo, useCallback } from 'react';
import ReactMapGL from 'react-map-gl';
import 'mapbox-gl/dist/mapbox-gl.css';

import Marker from './marker/Marker';

const MAP_SIZE = {
  height: '100vh',
  width: '100vw',
}

const INITIAL_VIEWPORT = {
  latitude: 51.40,
  longitude: 14.87,
  zoom: 12,
};

const MAPBOX_STYLE = 'mapbox://styles/medarilla/ckdvpnkyq13sg19qn16sah8fl';
const MAPBOX_ACCESS_TOKEN = process.env.REACT_APP_MAPBOX_ACCESS_TOKEN;

function Map({ columnsOrder, file }) {
  const [mapApi, setMapApi] = useState();
  const [viewport, setViewPort] = useState(INITIAL_VIEWPORT);

  const handleLoad = useCallback(
    ({ target: map }) => {
      setMapApi(map);
    },[],
  );

  const addressColumnIndex = columnsOrder.indexOf('address');
  const markers =  file.data.map((row, index) => {
    const [latitude, longitude] = file.coordinates[index];

    return (
      <Marker 
        key={`${row[addressColumnIndex]}-${index}`}
        address={row[addressColumnIndex]}
        latitude={latitude}
        longitude={longitude}
      />
    )
  });

  return (
    <ReactMapGL
      {...viewport}
      mapboxApiAccessToken={MAPBOX_ACCESS_TOKEN}
      mapStyle={MAPBOX_STYLE}
      onLoad={handleLoad}
      visible={Boolean(mapApi)}
      onViewportChange={setViewPort}
      {...MAP_SIZE}
    >
      {mapApi ? markers : <div>Spinner</div>}
    </ReactMapGL>
  );
}

export default memo(Map);
