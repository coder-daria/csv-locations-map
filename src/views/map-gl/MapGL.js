/* eslint-disable react/jsx-props-no-spreading */
import React, { useState, memo, useCallback } from 'react';
import { arrayOf, func, string, shape, number, instanceOf } from 'prop-types';

import ReactMapGL from 'react-map-gl';
import 'mapbox-gl/dist/mapbox-gl.css';

import MarkerLayer from './marker-layer/MarkerLayer';

import { Loading } from './loading';

import {
  INITIAL_VIEWPORT,
  MAP_STYLE,
  MAPBOX_ACCESS_TOKEN,
  MAPBOX_STYLE,
} from './constants';

import { Container, Button } from './MapGL.styles';

function MapGL({ file, resetData, columnsOrder }) {
  const [mapApi, setMapApi] = useState();
  const [viewport, setViewPort] = useState(INITIAL_VIEWPORT);

  const handleLoad = useCallback(({ target: map }) => {
    setMapApi(map);
  }, []);

  return (
    <Container>
      <ReactMapGL
        mapboxApiAccessToken={MAPBOX_ACCESS_TOKEN}
        mapStyle={MAPBOX_STYLE}
        onLoad={handleLoad}
        visible={Boolean(mapApi)}
        onViewportChange={setViewPort}
        {...viewport}
        {...MAP_STYLE}
      >
        {mapApi ? (
          <>
            <Button onClick={resetData}>Upload new file</Button>
            <MarkerLayer file={file} columnsOrder={columnsOrder} />
          </>
        ) : (
          <Loading />
        )}
      </ReactMapGL>
    </Container>
  );
}

MapGL.propTypes = {
  columnsOrder: arrayOf(string).isRequired,
  file: shape({
    coordinates: arrayOf(arrayOf(number)),
    data: arrayOf(arrayOf(string)),
    errors: instanceOf(Array),
    meta: instanceOf(Object),
    name: string,
  }).isRequired,
  resetData: func.isRequired,
};

export default memo(MapGL);
