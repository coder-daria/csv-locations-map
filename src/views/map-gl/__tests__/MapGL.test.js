import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { FILE } from '../../__mocks__/file.mock';

import MapGL from '../MapGL';

const setup = () => {
  const props = {
    file: FILE,
    columnsOrder: ['state', 'city', 'address', 'zip', 'category'],
  };

  // eslint-disable-next-line react/jsx-props-no-spreading
  return shallow(<MapGL {...props} />);
};

describe('<MapGL />', () => {
  it('renders without crashing', () => {
    const wrapper = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
