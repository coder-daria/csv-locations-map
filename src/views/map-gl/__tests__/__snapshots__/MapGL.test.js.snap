// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MapGL /> renders without crashing 1`] = `
<styled.div>
  <InteractiveMap
    altitude={1.5}
    asyncRender={false}
    attributionControl={true}
    bearing={0}
    className=""
    clickRadius={0}
    container={<body />}
    disableTokenWarning={false}
    doubleClickZoom={true}
    dragPan={true}
    dragRotate={true}
    getCursor={[Function]}
    height="100%"
    keyboard={true}
    latitude={47.51}
    longitude={16.79}
    mapOptions={Object {}}
    mapStyle="mapbox://styles/medarilla/ckdvpnkyq13sg19qn16sah8fl"
    mapboxApiAccessToken="pk.eyJ1IjoibWVkYXJpbGxhIiwiYSI6ImNrZHZvMDQ0cDBzY3YydHNncGlybG80cWoifQ.VIgYCyNdCDxhL7T1QNxSuQ"
    mapboxApiUrl="https://api.mapbox.com"
    maxPitch={60}
    maxZoom={24}
    minPitch={0}
    minZoom={2}
    onClick={null}
    onContextMenu={[Function]}
    onError={[Function]}
    onHover={null}
    onLoad={[Function]}
    onNativeClick={null}
    onResize={[Function]}
    onStateChange={[Function]}
    onTransitionEnd={[Function]}
    onTransitionInterrupt={[Function]}
    onTransitionStart={[Function]}
    onViewStateChange={null}
    onViewportChange={[Function]}
    pitch={0}
    preserveDrawingBuffer={false}
    preventStyleDiffing={false}
    reuseMaps={false}
    scrollZoom={true}
    style={null}
    touchAction="none"
    touchRotate={false}
    touchZoom={true}
    transitionDuration={0}
    transitionEasing={[Function]}
    transitionInterpolator={
      LinearInterpolator {
        "around": undefined,
        "propNames": Array [
          "longitude",
          "latitude",
          "zoom",
          "bearing",
          "pitch",
        ],
      }
    }
    transitionInterruption={1}
    visibilityConstraints={
      Object {
        "maxPitch": 60,
        "maxZoom": 24,
        "minPitch": 0,
        "minZoom": 0,
      }
    }
    visible={false}
    width="100%"
    zoom={2}
  >
    <Loading />
  </InteractiveMap>
</styled.div>
`;
